test/functional/core/startup_spec.lua

```diff
@@ -462,15 +462,23 @@ describe('startup', function()
   end)
 
   it('-r works without --headless in PTY #23294', function()
-    skip(is_os('win'))
-    eq({ 0 }, exec_lua([[return vim.fn.jobwait({ vim.fn.jobstart({...}, {
-      pty = true,
-      stdout_buffered = true,
-      on_stdout = function(_, data, _)
-        _G.Recovery_stdout = data
-      end,
-    }) })]], nvim_prog, '-u', 'NONE', '-i', 'NONE', '-r'))
-    matches('Swap files found:\r*', exec_lua('return _G.Recovery_stdout[1]'))
+    exec([[
+      func Normalize(data) abort
+        " Windows: remove ^M and term escape sequences
+        return map(a:data, 'substitute(substitute(v:val, "\r", "", "g"), "\x1b\\%(\\]\\d\\+;.\\{-}\x07\\|\\[.\\{-}[\x40-\x7E]\\)", "", "g")')
+      endfunc
+      func OnOutput(id, data, event) dict
+        let g:stdout = Normalize(a:data)
+      endfunc
+      call jobstart([v:progpath, '-u', 'NONE', '-i', 'NONE', '-r'], {
+      \ 'pty': v:true,
+      \ 'stdout_buffered': v:true,
+      \ 'on_stdout': function('OnOutput'),
+      \ })
+    ]])
+    retry(nil, nil, function()
+      eq('Swap files found:', eval('g:stdout[0]'))
```

Comment:
Off-topic: we should add a way for `nvim_exec2` to return a value from a script. Maybe by setting a `v:` var in the script, or introduce a `finish({var})` function.